File: core/src/main/java/com/ustadmobile/core/util/AsyncServiceManager.java
package com.ustadmobile.core.util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Vector;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Utility class to help manage a service that starts and stops asynchronously.
 *
 */
public abstract class AsyncServiceManager {

    public static final int STATE_STOPPED = 0;

    public static final int STATE_STARTING = 1;

    public static final int STATE_STARTED = 2;

    public static final int STATE_STOPPING = 3;

    private volatile int targetState;

    private volatile int currentState;

    private DelayedExecutor delayedExecutor;

    private final ReentrantLock lock = new ReentrantLock();

    private Vector<OnStateChangeListener> stateChangeListeners = new Vector<>();

    public interface DelayedExecutor {

        void runAfterDelay(Runnable runnable, long delay);

    }

    public interface OnStateChangeListener {
        void onStateChanged(AsyncServiceManager serviceManager, int newState);
    }

    public interface AsyncAwaitChecker {
        boolean stopWaiting(int newState);
    }

    public AsyncServiceManager(int initialState, DelayedExecutor delayedExecutor) {
        currentState = initialState;
        targetState = initialState;
        this.delayedExecutor = delayedExecutor;
    }

    //Blank constructor required for mocking for tests
    public AsyncServiceManager() {

    }

    protected void setDelayedExecutor(DelayedExecutor executor) {
        this.delayedExecutor = executor;
    }

    public void setEnabled(boolean enabled) {
        try {
            lock.lock();
            targetState = enabled ? STATE_STARTED : STATE_STOPPED;
            if(targetState == currentState
                    || targetState == STATE_STARTED && currentState == STATE_STARTING
                    || targetState == STATE_STOPPED && currentState == STATE_STOPPING) {
                //nothing to do
                return;
            }

            if(targetState == STATE_STARTED && currentState == STATE_STOPPED) {
                currentState = STATE_STARTING;
                fireStateChangedEvent(currentState);
                delayedExecutor.runAfterDelay(this::start, 0);
            }else if(targetState == STATE_STOPPED && currentState == STATE_STARTED) {
                currentState = STATE_STOPPING;
                fireStateChangedEvent(currentState);
                delayedExecutor.runAfterDelay(this::stop, 0);
            }else {
                delayedExecutor.runAfterDelay(this::checkState, 1000);
            }
        }finally {
            lock.unlock();
        }
    }

    private void checkState(){
        setEnabled(targetState == STATE_STARTED);
    }

    protected void notifyStateChanged(final int state, final int newTargetState) {
        try {
            lock.lock();
            this.currentState = state;

            if(newTargetState != -1)
                targetState = newTargetState;

        }finally {
            lock.unlock();
        }
        fireStateChangedEvent(state);
    }

    protected void notifyStateChanged(final int state) {
        notifyStateChanged(state, -1);
    }

    protected void fireStateChangedEvent(int newState) {
        ArrayList<OnStateChangeListener> listenerList = new ArrayList<>(stateChangeListeners.size());
        listenerList.addAll(stateChangeListeners);
        for(OnStateChangeListener listener : listenerList) {
            listener.onStateChanged(this, newState);
        }
    }

    public int getState() {
        return currentState;
    }

    public void addOnStateChangeListener(OnStateChangeListener listener) {
        stateChangeListeners.add(listener);
    }

    public void removeOnStateChangeListener(OnStateChangeListener listener) {
        stateChangeListeners.remove(listener);
    }

    public void await(AsyncAwaitChecker checker, long timeout, TimeUnit timeoutUnit) {
        if(checker.stopWaiting(currentState))
            return;

        CountDownLatch latch = new CountDownLatch(1);
        OnStateChangeListener listener = (service, newState) -> {
            if(checker.stopWaiting(newState))
                latch.countDown();
        };
        addOnStateChangeListener(listener);
        try { latch.await(timeout, timeoutUnit); }
        catch(InterruptedException e) { /*should not happen*/}
        removeOnStateChangeListener(listener);
    }


    public abstract void start();

    public abstract void stop();



}

File: core/src/main/java/com/ustadmobile/core/util/AsyncServiceManager.kt
package com.ustadmobile.core.util

import java.util.ArrayList
import java.util.Vector
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock

/**
 * Utility class to help manage a service that starts and stops asynchronously.
 *
 */
abstract class AsyncServiceManager {

    @Volatile
    private var targetState: Int = 0

    @Volatile
    var state: Int = 0
        private set

    private var delayedExecutor: DelayedExecutor? = null

    private val lock = ReentrantLock()

    private val stateChangeListeners = Vector<OnStateChangeListener>()

    interface DelayedExecutor {

        fun runAfterDelay(runnable: Runnable, delay: Long)

    }

    interface OnStateChangeListener {
        fun onStateChanged(serviceManager: AsyncServiceManager, newState: Int)
    }

    interface AsyncAwaitChecker {
        fun stopWaiting(newState: Int): Boolean
    }

    constructor(initialState: Int, delayedExecutor: DelayedExecutor) {
        state = initialState
        targetState = initialState
        this.delayedExecutor = delayedExecutor
    }

    //Blank constructor required for mocking for tests
    constructor() {

    }

    fun setDelayedExecutor(executor: DelayedExecutor) {
        this.delayedExecutor = executor
    }

    fun setEnabled(enabled: Boolean) {
        try {
            lock.lock()
            targetState = if (enabled) STATE_STARTED else STATE_STOPPED
            if (targetState == state
                    || targetState == STATE_STARTED && state == STATE_STARTING
                    || targetState == STATE_STOPPED && state == STATE_STOPPING) {
                //nothing to do
                return
            }

            if (targetState == STATE_STARTED && state == STATE_STOPPED) {
                state = STATE_STARTING
                fireStateChangedEvent(state)
                delayedExecutor!!.runAfterDelay(Runnable { this.start() }, 0)
            } else if (targetState == STATE_STOPPED && state == STATE_STARTED) {
                state = STATE_STOPPING
                fireStateChangedEvent(state)
                delayedExecutor!!.runAfterDelay(Runnable { this.stop() }, 0)
            } else {
                delayedExecutor!!.runAfterDelay(Runnable { this.checkState() }, 1000)
            }
        } finally {
            lock.unlock()
        }
    }

    private fun checkState() {
        setEnabled(targetState == STATE_STARTED)
    }

    @JvmOverloads
    fun notifyStateChanged(state: Int, newTargetState: Int = -1) {
        try {
            lock.lock()
            this.state = state

            if (newTargetState != -1)
                targetState = newTargetState

        } finally {
            lock.unlock()
        }
        fireStateChangedEvent(state)
    }

    protected fun fireStateChangedEvent(newState: Int) {
        val listenerList = ArrayList<OnStateChangeListener>(stateChangeListeners.size)
        listenerList.addAll(stateChangeListeners)
        for (listener in listenerList) {
            listener.onStateChanged(this, newState)
        }
    }

    fun addOnStateChangeListener(listener: OnStateChangeListener) {
        stateChangeListeners.add(listener)
    }

    fun removeOnStateChangeListener(listener: OnStateChangeListener) {
        stateChangeListeners.remove(listener)
    }

    fun await(checker: AsyncAwaitChecker, timeout: Long, timeoutUnit: TimeUnit) {
        if (checker.stopWaiting(state))
            return

        val latch = CountDownLatch(1)
        val listener = object : OnStateChangeListener {
            override fun onStateChanged(serviceManager: AsyncServiceManager, newState: Int) {
                if (checker.stopWaiting(newState))
                    latch.countDown()
            }
        }
        addOnStateChangeListener(listener)
        try {
            latch.await(timeout, timeoutUnit)
        } catch (e: InterruptedException) { /*should not happen*/
        }

        removeOnStateChangeListener(listener)
    }


    abstract fun start()

    abstract fun stop()

    companion object {

        val STATE_STOPPED = 0

        val STATE_STARTING = 1

        val STATE_STARTED = 2

        val STATE_STOPPING = 3
    }


}
